### ISSUE TITLE
Platinum analysis 

### ISSUE THREAD

**Issue by jyaacoub** [jyaacoub/MutDTA#94]
1. Build dataset
	1. [x] Copy all aflow platinum conf files from #narval to #h4h ✅ 2024-04-29
	2. [x] Init new platinum dataset with new confs ✅ 2024-04-29
```python
create_datasets(cfg.DATA_OPT.platinum, 
                cfg.PRO_FEAT_OPT.nomsa, cfg.PRO_EDGE_OPT.aflow, 
                ligand_features=cfg.LIG_FEAT_OPT.gvp,
                ligand_edges=cfg.LIG_EDGE_OPT.binary,
                k_folds=None, train_split=0, val_split=0)		
```
2. Run inference
	1. [x] Find trained weights for GVPL-aflow model on pdbbind (all 5 for each split to use as an ensemble) ✅ 2024-04-29
		- Model key is `GVPLM_PDBbind1D_nomsaF_aflowE_128B_0.00022659LR_0.02414D_2000E_gvpLF_binaryLE`
		- They are all on **#h4h**
	2. [x] Save predictions of each model to a csv file ✅ 2024-04-29
```python
loaders = Loader.load_DataLoaders(cfg.DATA_OPT.platinum,
                               cfg.PRO_FEAT_OPT.nomsa, cfg.PRO_EDGE_OPT.aflow,
                               ligand_feature=cfg.LIG_FEAT_OPT.gvp, ligand_edge=cfg.LIG_EDGE_OPT.binary,
                               datasets=['test'])

#%%
model = Loader.init_model(cfg.MODEL_OPT.GVPL, cfg.PRO_FEAT_OPT.nomsa, cfg.PRO_EDGE_OPT.aflow,
                          dropout=0.02414, output_dim=256)

#%%
cp_dir = "/cluster/home/t122995uhn/projects/MutDTA/results/model_checkpoints/ours"
MODEL_KEY = lambda fold: f"GVPLM_PDBbind{fold}D_nomsaF_aflowE_128B_0.00022659LR_0.02414D_2000E_gvpLF_binaryLE"
cp = lambda fold: f"{cp_dir}/{MODEL_KEY(fold)}.model"

out_dir = f'{cfg.MEDIA_SAVE_DIR}/test_set_pred/'
os.makedirs(out_dir, exist_ok=True)

for i in range(5):
    model.safe_load_state_dict(torch.load(cp(i), map_location=device))
    model.to(device)
    model.eval()

    loss, pred, actual = test(model, loaders['test'], device, verbose=True)
    
    # saving as csv with columns code, pred, actual
    # get codes from test loader
    codes, pid = [b['code'][0] for b in loaders['test']], [b['prot_id'][0] for b in loaders['test']]
    df = pd.DataFrame({'prot_id': pid, 'pred': pred, 'actual': actual}, index=codes)
    df.index.name = 'code'
    df.to_csv(f'{out_dir}/{MODEL_KEY(i)}_PLATINUM.csv')
```
4. Analysis  ^3fe12b
	1. [x] Check for overlap in pdbIDs and remove them 
		- provide a plot of with and without performance?
	2. how can I analyze the performance of the model in its ability to detect dangerous mutations?
		1. [x] Normal predictive performance analysis with cindex and MSE scores
			- Statistical t-test to determine how different the predicted and experimental distributions are from each other
		2. [x] Mutation impact analysis: 
			- calculate $\Delta pk_a$ (change in binding affinity for each protein-ligand pair). Then correlate the predicted $\Delta pk_a$ and the experimental $\Delta pk_a$ 
				- higher correlation indicates that the **model effectively captures the impact of mutations**
			- Split up scores depending on number of mutations for even more analysis (maybe it struggles with larger # of mutations)
		3. [x] Identifying "significant" mutations
			- Based on the distribution of $\Delta pk_a$ scores classify all above **[1 and 2] standard deviations to be "significant"**?
			- Then Build a confusion matrix to analyze the true positive rate and true negative rate of the model in identifying these mutations.

**Comment by jyaacoub** [comment]
### There is huge overlap between the PDBbind training data and the platinum dataset unfortunately...

Removing all exact instances of pdbids leaves us with 975 rows:
![image](https://github.com/jyaacoub/MutDTA/assets/50300488/9a5c2f50-84fc-40a4-bce6-8cca8041b6aa)

However, if we consider both mutated and wildtype proteins as the same protein then we are left with ZERO rows. 

The middle ground is to not remove wildtype proteins but this is still not optimal and leaves us with 1461 rows:
![image](https://github.com/jyaacoub/MutDTA/assets/50300488/68a1d728-ef9d-440d-969a-1c8edc60ff7e)

**Comment by jyaacoub** [comment]
## 1. predictive performance

| metric | With Overlap      | Without Overlap   | Significance |
| :----- | :---------------- | :---------------- | :----------- |
| mae    | 1.380 $\pm$ 0.059 | 1.467 $\pm$ 0.031 | *            |
| pcorr  | 0.530 $\pm$ 0.052 | 0.415 $\pm$ 0.064 | *            |
| cindex | 0.674 $\pm$ 0.023 | 0.641 $\pm$ 0.025 |              |
| mse    | 3.106 $\pm$ 0.272 | 3.467 $\pm$ 0.249 |              |
| scorr  | 0.503 $\pm$ 0.067 | 0.411 $\pm$ 0.086 |              |
| rmse   | 1.761 $\pm$ 0.078 | 1.861 $\pm$ 0.068 |              |


## CODE:

```python
from src.analysis.figures import fig_dpkd_dist, fig_sig_mutations_conf_matrix, generate_roc_curve
from src.analysis.metrics import get_metrics
import numpy as np
import pandas as pd
from scipy.stats import ttest_ind

MODEL = lambda i: f"results/model_media/test_set_pred/GVPLM_PDBbind{i}D_nomsaF_aflowE_128B_0.00022659LR_0.02414D_2000E_gvpLF_binaryLE_PLATINUM.csv" 

data_p = lambda set: f'/cluster/home/t122995uhn/projects/data/PDBbindDataset/nomsa_aflow_gvp_binary/{set}0/cleaned_XY.csv'
df_t = pd.Index.append(pd.read_csv(data_p('train'), index_col=0).index, 
                       pd.read_csv(data_p('val'), index_col=0).index)
df_t = df_t.str.upper()

results_with_overlap = []
results_without_overlap = []

for i in range(5):
    df = pd.read_csv(MODEL(i), index_col=0).dropna()
    df['pdb'] = df['prot_id'].str.split('_').str[0]

    # with overlap
    metrics_with = list(get_metrics(np.array(df['actual']), np.array(df['pred']), 
                               save_figs=False, save_data=False, show=False))
    # dont return p vals:
    metrics_with[1] = metrics_with[1][0]
    metrics_with[2] = metrics_with[2][0]
    results_with_overlap.append(metrics_with)

    # without overlap
    df_no_overlap = df[~(df['pdb'].isin(df_t))]
    metrics_without = list(get_metrics(np.array(df_no_overlap['actual']), np.array(df_no_overlap['pred']), 
                                  save_figs=False, save_data=False, show=False))
    metrics_without[1] = metrics_without[1][0]
    metrics_without[2] = metrics_without[2][0]
    results_without_overlap.append(metrics_without)

# Convert results to DataFrame
results_with_df = pd.DataFrame(results_with_overlap, columns=['cindex', 'pcorr', 'scorr', 'mse', 'mae', 'rmse'])
results_without_df = pd.DataFrame(results_without_overlap, columns=['cindex', 'pcorr', 'scorr', 'mse', 'mae', 'rmse'])

# Statistics computation
mean_with = results_with_df.mean()
std_with = results_with_df.std()
mean_without = results_without_df.mean()
std_without = results_without_df.std()

# T-tests for significance
ttests = {col: ttest_ind(results_with_df[col], results_without_df[col]) for col in results_with_df.columns}

# Combine mean and std
combined_with = mean_with.map(lambda x: f"{x:.3f}") + " +- " + std_with.map(lambda x: f"{x:.3f}")
combined_without = mean_without.map(lambda x: f"{x:.3f}") + " +- " + std_without.map(lambda x: f"{x:.3f}")

# Add significance marks
significance = pd.Series({col: '*' if ttests[col].pvalue < 0.05 else '' for col in results_with_df.columns})

# Prepare markdown table
md_table = pd.concat([combined_with, combined_without, significance], axis=1)
md_table.columns = ['With Overlap', 'Without Overlap', 'Significance']
md_output = md_table.to_markdown()
print(md_output)
```

**Comment by jyaacoub** [comment]
## 3. Significant Muation impact analysis

<details>
<summary>code </summary>
```python
from src.analysis.figures import fig_dpkd_dist, fig_sig_mutations_conf_matrix, generate_roc_curve
from src.analysis.metrics import get_metrics
import numpy as np
import pandas as pd
from scipy.stats import ttest_ind

MODEL = lambda i: f"results/model_media/test_set_pred/GVPLM_PDBbind{i}D_nomsaF_aflowE_128B_0.00022659LR_0.02414D_2000E_gvpLF_binaryLE_PLATINUM.csv" 

data_p = lambda set: f'/cluster/home/t122995uhn/projects/data/PDBbindDataset/nomsa_aflow_gvp_binary/{set}0/cleaned_XY.csv'
df_t = pd.Index.append(pd.read_csv(data_p('train'), index_col=0).index, 
                       pd.read_csv(data_p('val'), index_col=0).index)
df_t = df_t.str.upper()

results_with_overlap = []
results_without_overlap = []
i=0

df = pd.read_csv(MODEL(i), index_col=0).dropna()
df['pdb'] = df['prot_id'].str.split('_').str[0]
df_no_overlap = df[~(df['pdb'].isin(df_t))]

#%%
true_dpkd = fig_dpkd_dist(df, pkd_col='actual', verbose=True)
pred_dpkd = fig_dpkd_dist(df, pkd_col='pred', verbose=True)

#%%
conf1, tpr1, tnr1 = fig_sig_mutations_conf_matrix(true_dpkd, pred_dpkd, std=0.3)
conf2, tpr2, tnr2 = fig_sig_mutations_conf_matrix(true_dpkd, pred_dpkd, std=1)
# %%
# ROC 
generate_roc_curve(true_dpkd, pred_dpkd, thres_range=(0,5), step=0.1)
```
</details>

**Comment by jyaacoub** [comment]
# Pretrained Davis results
- [x] build dataset from aflow confirmations
- [x] Run hyperparameter tuning
- [ ] Train davis
- [ ] Run platinum evaluation
